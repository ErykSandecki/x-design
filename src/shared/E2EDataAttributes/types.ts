// types
import { E2EAttribute } from 'types/e2e';

export type TE2EType = E2EAttribute;
export type TE2EValue = number | string;
