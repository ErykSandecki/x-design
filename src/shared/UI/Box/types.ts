// types
import { THTMLTag } from 'types';

export type TBoxHTMLTag = Exclude<
  THTMLTag,
  | 'aduio'
  | 'base'
  | 'body'
  | 'br'
  | 'canvas'
  | 'col'
  | 'colgroup'
  | 'h1'
  | 'h2'
  | 'h3'
  | 'h4'
  | 'h5'
  | 'h6'
  | 'head'
  | 'html'
  | 'iframe'
  | 'img'
  | 'input'
  | 'meta'
  | 'noscript'
  | 'p'
  | 'param'
  | 'script'
  | 'small'
  | 'source'
  | 'style'
  | 'track'
>;
