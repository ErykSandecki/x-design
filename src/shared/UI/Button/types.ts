// types
import { TIconProps } from '../Icon/Icon';

export type TButtonIcon = null | Pick<TIconProps, 'name'>;

export type TButtonIconArgs = {
  placement: 'start' | 'end';
  src: TButtonIcon;
};
