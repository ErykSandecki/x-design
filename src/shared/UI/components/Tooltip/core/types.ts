export type TContext = {
  active: boolean;
  onMouseEnter: TFunc;
  onMouseLeave: TFunc;
};
