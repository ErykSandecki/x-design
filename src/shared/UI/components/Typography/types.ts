import { CSSProperties, JSX } from 'react';

// types
import { THTMLTag, TObject } from 'types';
import { TTypographyProps } from './Typography';
import { TypographyVariant } from './enums';

export type TOmittedTypographyProps = Omit<TTypographyProps, 'component' | 'fontType'> & {
  component?: TTypograpghyElement;
};

export type TTypograpghyComponentProps = TTypographyProps & {
  className: string;
  style: CSSProperties;
};

export type TTypograpghyElement = (props: TTypograpghyComponentProps) => JSX.Element;

export type TTypograpghyComponent = TTypograpghyElement | TTypograpghyHTMLTag;

export type TTypograpghyHTMLTag = Exclude<
  THTMLTag,
  | 'aduio'
  | 'base'
  | 'body'
  | 'br'
  | 'canvas'
  | 'col'
  | 'colgroup'
  | 'h1'
  | 'h2'
  | 'h3'
  | 'h4'
  | 'h5'
  | 'h6'
  | 'head'
  | 'html'
  | 'iframe'
  | 'img'
  | 'input'
  | 'meta'
  | 'noscript'
  | 'p'
  | 'param'
  | 'script'
  | 'small'
  | 'source'
  | 'style'
  | 'track'
>;

export type TTypograpghyVariantMapping = TObject<TTypograpghyHTMLTag, typeof TypographyVariant, 'optional'>;
