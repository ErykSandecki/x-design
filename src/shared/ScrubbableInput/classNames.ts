// utils
import { composeClassNames } from 'utils';

export const className = 'ScrubbableInput';

export const classNames = composeClassNames(className, [className, 'disabled'] as const, ['handler'] as const);
