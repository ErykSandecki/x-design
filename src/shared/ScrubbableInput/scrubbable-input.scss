@import '../../styles/z-indexes';

.ScrubbableInput {
  cursor: ew-resize !important;
  user-select: none;

  &--disabled {
    pointer-events: none;
  }

  &__handler {
    position: fixed;
    transform: translate(-50%, -7px);
    z-index: map.get($map: $--xd-z-indexes, $key: 'overlay');
  }
}
