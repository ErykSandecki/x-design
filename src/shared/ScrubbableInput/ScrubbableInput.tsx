import { createPortal } from 'react-dom';
import { FC, ReactNode, useRef } from 'react';
import { noop } from 'lodash';

// components
import E2EDataAttribute, { TE2EDataAttributeProps } from '../E2EDataAttributes/E2EDataAttribute';
import Box from '../UI/Box/Box';
import Icon, { TIconProps } from '../UI/Icon/Icon';

// hooks
import { useScrubbableInputEvents } from './hooks/useScrubbableInputEvents';
import { useTheme } from 'hooks';

// others
import { className, classNames } from './classNames';

// styles
import styles from './scrubbable-input.scss';

// types
import { ColorsTheme, E2EAttribute } from 'types';

export type TScrubbableInputProps = {
  children?: ReactNode;
  disabled?: boolean;
  e2eValue?: TE2EDataAttributeProps['value'];
  icon?: TIconProps['name'];
  iconHeight?: number;
  iconWidth?: number;
  loop?: boolean;
  max: number;
  min: number;
  onChange: TFunc<[number]>;
  onMouseDown?: TFunc;
  onMouseUp?: TFunc;
  value: number;
};

export const ScrubbableInput: FC<TScrubbableInputProps> = ({
  children,
  disabled = false,
  e2eValue = '',
  icon,
  iconHeight = 12,
  iconWidth = 12,
  loop,
  max,
  min,
  onChange,
  onMouseDown = noop,
  onMouseUp = noop,
  value,
}) => {
  const inputRef = useRef(null);
  const { classNamesWithTheme, cx } = useTheme(classNames, styles);

  const { mousePosition, ...events } = useScrubbableInputEvents(
    inputRef,
    loop,
    max,
    min,
    onChange,
    onMouseDown,
    onMouseUp,
    value,
  );

  return (
    <E2EDataAttribute type={E2EAttribute.scrubbableInput} value={e2eValue}>
      <Box
        classes={{
          className: cx(classNamesWithTheme[className].name, [
            classNamesWithTheme[className].modificators.disabled,
            disabled,
          ]),
        }}
        ref={inputRef}
        sx={{ alignItems: 'center', display: 'flex' }}
        {...events}
      >
        {children}
        {icon && (
          <Icon color={ColorsTheme.neutral2} disabled={disabled} height={iconHeight} name={icon} width={iconWidth} />
        )}
        {mousePosition &&
          createPortal(
            <Icon
              classes={{ className: cx(classNamesWithTheme.handler) }}
              height={14}
              name="EWResize"
              style={{
                left: `${mousePosition.x}px`,
                top: `${mousePosition.y}px`,
              }}
              width={25}
            />,
            document.body,
          )}
      </Box>
    </E2EDataAttribute>
  );
};

export default ScrubbableInput;
