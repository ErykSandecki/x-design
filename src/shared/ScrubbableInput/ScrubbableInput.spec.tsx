import { fireEvent } from '@testing-library/react';

// components
import ScrubbableInput from './ScrubbableInput';

// types
import { E2EAttribute } from 'types/e2e';

// utils
import { customRender, getByE2EAttribute } from 'test';

const mockCallBack = jest.fn();

describe('ScrubbableInput snapshots', () => {
  it('should render ScrubbableInput', () => {
    // before
    const { asFragment } = customRender(
      <ScrubbableInput max={100} min={0} onChange={mockCallBack} value={0}>
        <span>child</span>
      </ScrubbableInput>,
    );

    // result
    expect(asFragment()).toMatchSnapshot();
  });

  it('should render with icon', () => {
    // before
    const { asFragment, container } = customRender(
      <ScrubbableInput max={100} min={0} onChange={mockCallBack} value={0}>
        <span>child</span>
      </ScrubbableInput>,
    );

    // action
    fireEvent.mouseDown(getByE2EAttribute(container, E2EAttribute.scrubbableInput));

    // result
    expect(asFragment()).toMatchSnapshot();
  });
});
