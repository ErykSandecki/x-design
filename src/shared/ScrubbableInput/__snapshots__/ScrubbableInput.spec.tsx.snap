// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrubbableInput snapshots should render ScrubbableInput 1`] = `
<DocumentFragment>
  <div
    class="ScrubbableInput ScrubbableInput--light"
    data-test-scrubbable-input=""
  >
    <span>
      child
    </span>
  </div>
</DocumentFragment>
`;

exports[`ScrubbableInput snapshots should render with icon 1`] = `
<DocumentFragment>
  <div
    class="ScrubbableInput ScrubbableInput--light"
    data-test-scrubbable-input=""
  >
    <span>
      child
    </span>
    <div
      class="css-1jnxvbp ScrubbableInput__handler ScrubbableInput__handler--light Icon Icon--light Icon--neutral1 Icon--neutral1--light"
      data-test-icon="ew-resize"
      height="14"
      name="EWResize"
      style="left: 0px; top: 0px;"
      width="25"
    />
  </div>
</DocumentFragment>
`;
